Source/Core/VideoBackends/Software/TransformUnit.cpp

```diff
@@ -90,22 +90,14 @@ void TransformPosition(const InputVertexData* src, OutputVertexData* dst)
   }
 }
 
-void TransformNormal(const InputVertexData* src, bool nbt, OutputVertexData* dst)
+void TransformNormal(const InputVertexData* src, OutputVertexData* dst)
 {
   const float* mat = &xfmem.normalMatrices[(src->posMtx & 31) * 3];
 
-  if (nbt)
-  {
-    MultiplyVec3Mat33(src->normal[0], mat, dst->normal[0]);
-    MultiplyVec3Mat33(src->normal[1], mat, dst->normal[1]);
-    MultiplyVec3Mat33(src->normal[2], mat, dst->normal[2]);
-    dst->normal[0].Normalize();
-  }
-  else
-  {
-    MultiplyVec3Mat33(src->normal[0], mat, dst->normal[0]);
-    dst->normal[0].Normalize();
-  }
+  MultiplyVec3Mat33(src->normal[0], mat, dst->normal[0]);
+  MultiplyVec3Mat33(src->normal[1], mat, dst->normal[1]);
+  MultiplyVec3Mat33(src->normal[2], mat, dst->normal[2]);
```

Comment:
It would be interesting to test what happens when using a regular texgen with the source row as BinormalT or BinormalN.  Only the regular texgen allows for use of one or two texture coordinate matrices, so maybe the time that would be spent for that is used to transform the normals otherwise?  (I'm not sure what would be happening with Color0 or Color1 though, as I'd assume the color would already be transformed by that point so I don't know what extra work it would be doing)

Reply:
Well when regular texgen use binormals as source, they get the raw binormals without any transform. 

It's also worth noting that diagrams in the patents show separate functional units for applying position/normal transforms (dot product unit) and texgen's texture coordinate matrix operations (texture dot 2; Do texture coordinate matrix operations happen in fixed point?), so as long as there are a few texgen operations before any emboss map operations, the binomals transform would have time to complete in parallel

(And then lighting is done in a completely separate pipeline with it's own functional units)

![image](https://user-images.githubusercontent.com/138484/164121231-e12afe16-a67b-4ccf-b5bd-cef81c5ed0fa.png)

Reply:
It's also odd that [the article](https://www.gamedeveloper.com/programming/shader-integration-merging-shading-technologies-on-the-nintendo-gamecube) says:

> Finally, emboss mapping (as bump mapping) needs binormals to describe the orientation of the height field on the surface. Since they need to be transformed the same way the normals are transformed this can add a bit overhead.

and also

> This means that only one vertex format is needed for the entire landscape, and it saves memory, transfer bandwidth and most importantly transform performance. 

That implies that it _is_ faster (and they could have measured the performance improvement with `GX_PERF0_XF_XFRM_CLKS`, and it's likely they did).  The article doesn't mention changing the normal matrix, though, and says that "XF will provide the previously transformed bi-normal and tangent, which reside in internal registers", which is inconsistent with what we've determined.  (The article was written for RS2, which doesn't change the normal matrix while using this trick (that I've seen), so this may just have been their understanding of how it worked at the time.)

Reply:
I really suspect faulty-recall here. That they forgot exactly what they were optimising for between when they did the optimisation and when they wrote the article. 

It doesn't seem like they were actually optimising for `GX_PERF0_XF_XFRM_CLKS`. If I'm understanding the graphical effect correctly, the amount of emboss is identical across the entire frame.   
It feels like there should have been multiple ways to achieve the same effect that would have been cheaper for both XF and TEV performance, at the expense of some per-frame overhead.  For example, baking the summed emboss effect into a texture each frame would have reduced the number of texture lookups by one, potentially eliminated a TEV stage and or or two XF texgen channels. 

With the fact that we know the transform gets applied anyway, and with the wild overlapping of color channels, it really feels like they are optimising solely for memory bandwidth. Both CPU -> RAM bandwidth, and RAM -> CP bandwidth.